### PR TITLE
Add driver for quad SPI AMOLED displays

### PR DESCRIPTION
**Pull request by clydebarrow**
# What does this implement/fix?

This PR implements a driver for quad SPI AMOLED displays as used in the Lilygo T-Display S3 AMOLED and T4-S3 boards.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#3677

## Test Environment

- [ ] ESP32
- [x] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
spi:
  id: quad_spi
  type: quad
  clk_pin: 15
  data_pins: [14, 10, 16, 12]

display:
  - platform: qspi_amoled
    model: RM690B0
    data_rate: 80MHz
    spi_mode: mode0
    dimensions:
      width: 450
      height: 600
      offset_width: 16
    color_order: rgb
    invert_colors: false
    brightness: 255
    cs_pin: 11
    reset_pin: 13
    enable_pin: 9

  - platform: qspi_amoled
    model: RM67162
    id: main_lcd
    dimensions:
      height: 240
      width: 536
    transform:
      mirror_x: true
      swap_xy: true
    color_order: rgb
    brightness: 255
    cs_pin: 6
    reset_pin: 17
    enable_pin: 38
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by clydebarrow** [comment]
Note - this will require a minor change (update test file) once #6353 is merged. So delay merging this until that one is done.

**Comment by clydebarrow** [comment]
Ready to go now.

File: esphome/components/qspi_amoled/display.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+                cv.Optional(CONF_TRANSFORM): cv.Schema(
+                    {
+                        cv.Optional(CONF_MIRROR_X, default=False): cv.boolean,
+                        cv.Optional(CONF_MIRROR_Y, default=False): cv.boolean,
+                        cv.Optional(CONF_SWAP_XY, default=False): cv.boolean,
+                    }
+                ),
+                cv.Optional(CONF_COLOR_ORDER, default="RGB"): cv.one_of(
+                    *COLOR_ORDERS.keys(), upper=True
+                ),
```
Comment (jesserockz) [inline esphome/components/qspi_amoled/display.py]:
```suggestion
                cv.Optional(CONF_COLOR_ORDER, default="RGB"): cv.enum(
                    COLOR_ORDERS, upper=True
                ),
```

File: esphome/components/qspi_amoled/display.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+    cv.only_with_esp_idf,
+)
+
+
+async def to_code(config):
+    var = cg.new_Pvariable(config[CONF_ID])
+    await display.register_display(var, config)
+    await spi.register_spi_device(var, config)
+
+    cg.add(var.set_color_mode(COLOR_ORDERS[config[CONF_COLOR_ORDER]]))
```
Comment (jesserockz) [inline esphome/components/qspi_amoled/display.py]:
```suggestion
    cg.add(var.set_color_mode(config[CONF_COLOR_ORDER]))
```

File: esphome/components/qspi_amoled/qspi_amoled.cpp [hunk 3]
```diff
@@ -0,0 +1,168 @@
+//
+// Created by Clyde Stubbs on 29/10/2023.
+//
+#include "qspi_amoled.h"
+#include "esphome/core/log.h"
+
+namespace esphome {
+namespace qspi_amoled {
+
+#ifdef USE_ESP_IDF
```
Comment (jesserockz) [inline esphome/components/qspi_amoled/qspi_amoled.cpp]: Generally try to put this as high in the file as possible

**Review by jesserockz — CHANGES_REQUESTED** [review]

File: esphome/components/qspi_amoled/qspi_amoled.cpp (hunk 3 above)
Comment (clydebarrow) [reply esphome/components/qspi_amoled/qspi_amoled.cpp]: How much higher can it go? My experience is that a namespace is required, or CI complains.

Comment (clydebarrow) [reply esphome/components/qspi_amoled/qspi_amoled.cpp]: Ok, I see it at the top elsewhere.

**Review by jesserockz — DISMISSED** [review]

**Review by jesserockz — APPROVED** [review]

**Pull request merged**